client/network/src/protocol/sync.rs

```diff
@@ -1156,60 +1195,119 @@ impl<B: BlockT> ChainSync<B> {
 		self.pending_requests.set_all();
 	}
 
-	/// Call when a node announces a new block.
+	/// Pre-validate the given block announcement.
 	///
-	/// If `OnBlockAnnounce::ImportHeader` is returned, then the caller MUST try to import passed
-	/// header (call `on_block_data`). The network request isn't sent
-	/// in this case. Both hash and header is passed as an optimization
-	/// to avoid rehashing the header.
-	pub fn on_block_announce(&mut self, who: &PeerId, hash: B::Hash, announce: &BlockAnnounce<B::Header>, is_best: bool)
-		-> OnBlockAnnounce
-	{
+	/// The pre-validation will return a [`Future`] that resolves to a
+	/// [`PreValidateBlockAnnounce`]. This result needs to be feed into
+	/// [`ChainSync::process_block_announce_pre_validation`] to finish
+	/// the block announcement processing.
+	pub fn pre_validate_block_announce(
+		&mut self,
+		who: PeerId,
+		hash: &B::Hash,
+		announce: BlockAnnounce<B::Header>,
+		is_best: bool,
+	) -> FEither<
+		impl Future<Output = PreValidateBlockAnnounce<B::Header>> + Send + Unpin,
+		impl Future<Output = PreValidateBlockAnnounce<B::Header>> + Send + Unpin
+	> {
```

Comment:
This can be an async function.

```suggestion
	pub async fn pre_validate_block_announce(
		&mut self,
		who: PeerId,
		hash: &B::Hash,
		announce: BlockAnnounce<B::Header>,
		is_best: bool,
	) -> PreValidateBlockAnnounce<B::Header> {
```